### ISSUE TITLE
Concurrent pulls fail

### ISSUE THREAD

**Issue by jhspaybar** [docker/docker#9718]
This issue seems to be triggered while pulling multiple images which map up to an identical base image.  A reproducible example is below.  This has been reproduced on Ubuntu 14.04 with DeviceMapper and the Amazon ECS-Optimized Amazon Linux AMI.  This has also been reproduced on Docker 1.3.3 and Docker 1.4.1.
```
    [ec2-user@ip-172-31-35-232 ~]$ docker version
    Client version: 1.4.1
    Client API version: 1.16
    Go version (client): go1.3.3
    Git commit (client): 5bc2ff8
    OS/Arch (client): linux/amd64
    Server version: 1.4.1
    Server API version: 1.16
    Go version (server): go1.3.3
    Git commit (server): 5bc2ff8
    [ec2-user@ip-172-31-35-232 ~]$ 
    [ec2-user@ip-172-31-35-232 ~]$ docker -D info
    Containers: 0
    Images: 0
    Storage Driver: devicemapper
    Pool Name: docker-202:1-1835239-pool
    Pool Blocksize: 65.54 kB
    Data file: /var/lib/docker/devicemapper/devicemapper/data
    Metadata file: /var/lib/docker/devicemapper/devicemapper/metadata
    Data Space Used: 307.2 MB
    Data Space Total: 107.4 GB
    Metadata Space Used: 725 kB
    Metadata Space Total: 2.147 GB
    Library Version: 1.02.82-git (2013-10-04)
    Execution Driver: native-0.2
    Kernel Version: 3.14.26-24.46.amzn1.x86_64
    Operating System: Amazon Linux AMI 2014.09
    CPUs: 1
    Total Memory: 1.957 GiB
    Name: ip-172-31-35-232
    ID: 2RCA:OSRS:4M6S:XHJG:MAUC:IYAA:KIGE:K2HM:VBRF:SCXN:LTRD:T3WP
    Debug mode (server): false
    Debug mode (client): true
    Fds: 57
    Goroutines: 58
    EventsListeners: 0
    Init Path: /usr/bin/docker
    Docker Root Dir: /var/lib/docker
    [ec2-user@ip-172-31-35-232 ~]$ uname -a
    Linux ip-172-31-35-232 3.14.26-24.46.amzn1.x86_64 #1 SMP Wed Dec 10 10:02:43 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
    [ec2-user@ip-172-31-35-232 ~]$ docker images
    REPOSITORY TAG IMAGE ID CREATED VIRTUAL SIZE
    [ec2-user@ip-172-31-35-232 ~]$ cat repro
    docker run nginx &
    docker run mysql &
    [ec2-user@ip-172-31-35-232 ~]$ sh repro
    [ec2-user@ip-172-31-35-232 ~]$ Unable to find image 'mysql:latest' locally
    Unable to find image 'nginx:latest' locally
    mysql:latest: The image you are pulling has been verified
     
    f10807909bc5: Pulling fs layer
    f6fab3b798be: Pulling fs layer
    d832c6f40cc3: Pulling fs layer
    cfab4f0d8972: Pulling fs layer
    cf8221608a63: Pulling fs layer
    4755e012c4cf: Pulling fs layer
    e012a865bac8: Pulling fs layer
    e2778dfe9ae9: Pulling fs layer
    b15206a696ec: Pulling fs layer
    9280b0f1af9a: Pulling fs layer
    9f1c487e78cd: Pulling fs layer
    10491040baf9: Pulling fs layer
    d8978d323757: Pulling fs layer
    f10807909bc5: Download complete
    f10807909bc5: Extracting [=================================================> ] 89.69 MB/90.17 MB 0
    d832c6f40cc3: Download complete
    f10807909bc5: Pull complete
    f10807909bc5: Download complete
    f6fab3b798be: Error downloading dependent layers
    d21beea329f5: Download complete
    04499cf33a0e: Download complete
    b15206a696ec: Downloading [====> ] 10.8 MB/118.7 MB 19s
    b15206a696ec: Download complete
    92360dceda1a: Download complete
    9495a555677d: Download complete
    911e2df03930: Download complete
    88daddf047c6: Download complete
    db8a64f62509: Download complete
    e46b3488b010: Download complete
    e46b3488b010: Error pulling image (latest) from nginx, Driver devicemapper failed to create image rootfs f10807909bc552de261ca7463effc467600c3dca68d8e7704425283a6911d2ca: device f10807909bc552de261ca7463effc467600c3dca68d8e7704425283a6911d2ca already exists 3dca68d8e7704425283a6911d2ca already exists
    FATA[0032] Error pulling image (latest) from nginx, Driver devicemapper failed to create image rootfs f10807909bc552de261ca7463effc467600c3dca68d8e7704425283a6911d2ca: device f10807909bc552dea385835786ac: Error pulling image (latest) from mysql, Error mounting '/dev/mapper/docker-202:1-1835239-f6fab3b798be3174f45aa1eb731f8182705555f89c9026d8c1ef230cbf8301dd' on '/var/lib/docker/devicemapper/mnt/f6fab3b798be3174f45aa1eb731f8182705555f89c9026d8c1ef230cbf8301dd': invalid argument 9c9026d8c1ef230cbf8301dd': invalid argument
    FATA[0034] Error pulling image (latest) from mysql, Error mounting '/dev/mapper/docker-202:1-1835239-f6fab3b798be3174f45aa1eb731f8182705555f89c9026d8c1ef230cbf8301dd' on '/var/lib/docker/devicemapper/mnt/f6fab3b798be3174f45aa1eb731f8182705555f89c9026d8c1ef230cbf8301dd': invalid argument
    ^C
    [ec2-user@ip-172-31-35-232 ~]$ docker ps -a
    CONTAINER ID IMAGE COMMAND CREATED STATUS PORTS NAMES
    [ec2-user@ip-172-31-35-232 ~]$ docker images
    REPOSITORY TAG IMAGE ID CREATED VIRTUAL SIZE
    <none> <none> f10807909bc5 5 weeks ago 85.1 MB
```

**Comment by rhvgoyal** [comment]
I think this issue is not related to devicemapper. The crux of the issue seems to be that there is a common layer of fs to be downloaded between two images (nignx and mysql). In my tests that layer is "1aeada447715". In latest docker I can see that first nginx instance downloads and extracts that layer and then docker mysql tries to download and extract same layer and then it finds that directory is not
empty and it fails.

Error pulling image (latest) from mysql, rename /var/lib/docker/graph/_tmp/31889e84f5404c06a1b73c39bf57018597a749360999e03a7b17a57978c4237f /var/lib/docker/graph/1aeada4477158496dc31ee5c6e7174240140d83fddf94bc57fc02bee1b04e44f: directory not empty 

If two docker instances are run one after another, then second one somehow recognizes that
image layer has already been downloaded and extracted and it does not have to do it again and 
things work.

So docker core needs to fix this so that two instances don't try to extract same layer twice and run
into various kind of issues as that layer has already been extracted.

**Comment by rhvgoyal** [comment]
One can try running two separate commands in separate terminals to clearly see that a common fs layer is being downloaded and extracted twice.

**Comment by vbatts** [comment]
Okay. if this issue is not specific to a graphdriver, but a race in `./graph/` then lets get some tests written for exposing this.

**Comment by termie** [comment]
-subscribe-

**Comment by JeanMertz** [comment]
It has been pretty quiet in this issue for some time now. However, when booting up a fresh CoreOS instance, and pulling/starting several (5+) images at the same time, I am still seeing issues.

A snippet from our logs:

```
May 14 08:17:06 manager-01 dockerd[967]: time="2015-05-14T08:17:06Z" level=error msg="Error from V2 registry: Image 511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158 already exists"
May 14 08:17:06 manager-01 dockerd[967]: Error pulling image (latest) from gliderlabs/registrator, Driver overlay failed to create image rootfs 46e263e5de56e0bb41d61486776d70185b196d792cbbc1b48ab30effb04e882d: lstat /var/lib/docker/overlay/511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158: no such file or directory
May 14 08:17:06 manager-01 dockerd[967]: time="2015-05-14T08:17:06Z" level=error msg="Error from V2 registry: Driver overlay failed to create image rootfs 46e263e5de56e0bb41d61486776d70185b196d792cbbc1b48ab30effb04e882d: lstat /var/lib/docker/overlay/511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158: no such file or directory"
May 14 08:17:09 manager-01 dockerd[967]: Error pulling image (latest) from blendle/aws-cli, Driver overlay failed to create image rootfs a8be93256def01920787a50e83e749075ffe63596c36fdd01885f4826074e4dd: lstat /var/lib/docker/overlay/511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158: no such file or directory
May 14 08:17:10 manager-01 dockerd[967]: Error pulling image (latest) from progrium/consul, Driver overlay failed to create image rootfs 46e263e5de56e0bb41d61486776d70185b196d792cbbc1b48ab30effb04e882d: lstat /var/lib/docker/overlay/511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158: no such file or directory
May 14 08:17:11 manager-01 dockerd[967]: Error pulling image (latest) from gliderlabs/registrator, Driver overlay failed to create image rootfs 46e263e5de56e0bb41d61486776d70185b196d792cbbc1b48ab30effb04e882d: lstat /var/lib/docker/overlay/511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158: no such file or directory
May 14 08:17:17 manager-01 dockerd[967]: Error pulling image (latest) from gliderlabs/registrator, Driver overlay failed to create image rootfs 46e263e5de56e0bb41d61486776d70185b196d792cbbc1b48ab30effb04e882d: lstat /var/lib/docker/overlay/511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158: no such file or directory
May 14 09:48:18 manager-01 dockerd[967]: Error pulling image (latest) from gliderlabs/registrator, Driver overlay failed to create image rootfs 46e263e5de56e0bb41d61486776d70185b196d792cbbc1b48ab30effb04e882d: lstat /var/lib/docker/overlay/511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158: no such file or directory
May 14 09:48:33 manager-01 dockerd[967]: Error pulling image (latest) from gliderlabs/registrator, Driver overlay failed to create image rootfs 46e263e5de56e0bb41d61486776d70185b196d792cbbc1b48ab30effb04e882d: lstat /var/lib/docker/overlay/511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158: no such file or directory
```

**Comment by alicegklein** [comment]
On 2 machines when I run
$ docker run -it ubuntu bash 
hangs pulling fs layer

The last part of of the run output:
07f8e8c5e660: Download complete
07f8e8c5e660: Download complete
07f8e8c5e660: Verifying Checksum
07f8e8c5e660: Downloading     32 B/32 B
07f8e8c5e660: Layer already being pulled by another client. Waiting.
07f8e8c5e660: Pulling fs layer
latest: Pulling from ubuntu
07f8e8c5e660: Pulling fs layer

**Comment by gregarndt** [comment]
This seems to be somewhat related to the original issue described, although not specific to devicemapper.  We have been using 1.6.1 in production for a couple of weeks and did not hit an issue until we moved to the new registry at registry.hub.docker.com (we were previously using quay.io on v1).

When a machine attempts to make multiple concurrent pulls for the image, we see this on some instances:

64dc4f9ad22a - Error pulling image (0.3.3) from taskcluster/tester, endpoint: https://registry-1.docker.io/v1/, Driver aufs failed to create image rootfs a5208e8002346d5588c27b5a0d6b49c0735e44ccb776bd746b23b67af3cf2440: open /mnt/var/lib/docker/aufs/layers/93c381d2c25533cc6fcf6ea8d208745601b3cd07d43209a3b7d1b09c58d634af: no such file or directory

This leaves an untagged image (<none>) listed and the image can never be downloaded unless a docker restart happens.  I've also noticed that while this image is pulling down, there is already a container running that must share a layer that the above image also uses because when attempting to remove the untagged image I get a warning that it can't be removed because a container is using some layer it needs to remove.  The other container is up and running fine though.

If this should be moved to a new issue let me know or if there is some other information I could provide I would be happy to do so.

These are ec2 nodes where we're seeing the issue.

$ docker info
Containers: 3
Images: 16
Storage Driver: aufs
 Root Dir: /mnt/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 21
 Dirperm1 Supported: false
Execution Driver: native-0.2
Kernel Version: 3.13.0-48-generic
Operating System: Ubuntu 14.04.2 LTS
CPUs: 4
Total Memory: 14.69 GiB

$ docker version
Client version: 1.6.1
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 97cd073
OS/Arch (client): linux/amd64
Server version: 1.6.1
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 97cd073
OS/Arch (server): linux/amd64

$ cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=14.04
DISTRIB_CODENAME=trusty
DISTRIB_DESCRIPTION="Ubuntu 14.04.2 LTS"

**Comment by thaJeztah** [comment]
@vbatts @stevvooe perhaps you can have a look here;

> and did not hit an issue until we moved to the new registry at registry.hub.docker.com (we were previously using quay.io on v1).

I think this will be because pulling from the new registry will pull up to 5 layers in parallel; an issue that was not so common may become a lot more relevant because of that?

**Comment by stevvooe** [comment]
@thaJeztah Is it possible there is a bookkeeping mistake in the layer references for the downloaded layer that has encountered an error? If you kill the other container referencing that image, does it allow the pull?

**Comment by gregarndt** [comment]
Yup, if the other container is killed (which was created from a different image), the original image that had issues can be pulled.

**Comment by alicegklein** [comment]
Thanks Greg,  Yes it now completes.

**Comment by termie** [comment]
this is becoming an increasingly problematic issue for my company, is there a workaround that will let me specify my own files rather than going through the docker graph store?

**Comment by msiedlarek** [comment]
+1, Docker v1.6.2 with the new registry v2.0.1
This issue basically breaks our deployment and, I would think, any serious Docker deployment beyond a single-container to-do app.

**Comment by vbatts** [comment]
i'm struggling to reproduce this. It it not a devicemapper specific issue, but likely a racy, lack-of-lock issue with storage drivers in general. I'm trying pulling concurrent with same name and different named images that reference same layer IDs, and consistently see the message that the image or layer is already being fetched.

Could you provide a concise reproduction steps?

**Comment by gregarndt** [comment]
@vbatts I'll attempt to have a more reproducible set of steps.  So far we're only seeing this on our AWS ec2 nodes, but not every pull.  So far the only thing I've noticed that works is restarting docker.

**Comment by jfrazelle** [comment]
here is a stack  https://paste.j3ss.co/zKmtbVXU/raw

**Comment by grier** [comment]
+1

With 1.6.2 and registry v2 this happens often enough that we no longer allow multiple container pulls per node now.

**Comment by jbrook** [comment]
Can anyone suggest a workaround please? This is happening to after the launch of 3 AWS instances when then each instance pulls the registry 2.0 image. Is it possible to disable parallel pull?

**Comment by jnsaff** [comment]
+1 having the same problem.

**Comment by stevvooe** [comment]
@jbrook @jnsaff @grier I understand this is rather frustrating. I have put together a few attempts that should single track these downloads with no effect. I don't have answers for you yet, but I want you to know we are looking into this.

The main issue is that we don't have a rock solid workaround, since we don't have a repeatable reproduction. A reproduction from a fresh docker install would go a long way here to get this resolved. Without a reproduction, we are left with code inspection to see if we can identify a root cause. Code inspection is obviously much slower.

@jfrazelle was able to reproduce the issue. It went away after removing `/var/lib/docker` then single tracking the downloads up until the affected layer. I am only able to reproduce this intermittently _with certain images_. My current working theory is that it happens when concurrently downloading images based on a common root layer. Also, I am nearly certain that is has something to do with pulling one image via v1 and one via v2 or vice versa, rather than having this be a pure v2 problem.

Other tasks have me researching this code in further detail. Usually a good audit of functionality will suss these kinds of problems right out. I'll keep everyone informed of the progress here.

**Comment by aaron42net** [comment]
@jbrook I use CoreOS and wrapped this flock around the docker pulls in my Fleet units, to limit docker to one concurrent pull at a time:

```
ExecStartPre=-/usr/bin/flock /home/core/pull.lock /usr/bin/timeout 5m /usr/bin/docker pull ...
```

**Comment by jnsaff** [comment]
Images that are good for reproducing this are:
mesosphere/chronos:chronos-2.3.4-1.0.81.ubuntu1404-mesos-0.22.1-1.0.ubuntu1404
mesosphere/mesos-master:0.22.1-1.0.ubuntu1404
mesosphere/mesos-slave:0.22.1-1.0.ubuntu1404
mesosphere/marathon:v0.8.2-RC4

They are all pretty large and stem from common ubuntu ancestor and the mesos ones share more layers.

For me the problem manifests when using upstart to start services concurrently. The workaround seems to emit events so upstart knows to back off.

**Comment by gregarndt** [comment]
We start a process using upstart that waits for docker to be ready and then will do concurrent pulls.  I have noticed that the machines having issues have a container already started that shares a common ancestor layer.

**Comment by stevvooe** [comment]
@jnsaff Interesting. Investigation shows that those particular images are not available via the V2 registry. None of those tags show up, but I can pull them fine with the docker client. @dmp42

**Comment by andytinycat** [comment]
+1, this is a huge bug. Hard to reproduce, but at high concurrency (30+ processes trying to pull the *same* container) I can get it to happen. I've had to resort to shimming Docker with my own shell script to call `flock` before doing a pull.

**Comment by thaJeztah** [comment]
@stevvooe should this be moved to the 1.7.1 milestone as well? Or would that not be realistic (I know it's a hard one)

**Comment by stevvooe** [comment]
@thaJeztah This has already been fixed by #14193. It is the same issue.

**Comment by thaJeztah** [comment]
Thanks @stevvooe, I thought it was. For those interested and willing to test (please!), the 1.7.1 release is currently being prepared here: https://github.com/docker/docker/pull/14264

Release candidates should be available soon, so helping to these this would be much appreciated :+1: 

I'm closing this issue, because it should be resolved by https://github.com/docker/docker/pull/14193

**Comment by zanella** [comment]
[root@kubernetes ~]# docker pull centos:7
7: Pulling from centos
f1b10cd84249: Download complete 
f1b10cd84249: Error downloading dependent layers 
c852f6d61e65: Downloading [=====================>                             ] 26.47 MB/62.89 MB
7322fbe74aa5: Error pulling image (7) from centos, endpoint: https://registry-1.docker.io/v1/, Driver devicemapper f7322fbe74aa5: Error pulling image (7) from centos, Driver devicemapper failed to create image rootfs f1b10cd842498c2Error pulling image (7) from centos, Driver devicemapper failed to create image rootfs f1b10cd842498c23d206ee0cbeaa9de8d2ae09ff3c7af2723a9e337a6965d639: Error running DeviceCreate (createSnapDevice) dm_task_run failed


[root@kubernetes ~]# docker version
Client version: 1.7.1
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 786b29d
OS/Arch (client): linux/amd64
Server version: 1.7.1
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 786b29d
OS/Arch (server): linux/amd64
[root@kubernetes ~]#

**Comment by stevvooe** [comment]
@zanella I am not sure what you want by just posting random output without any context. Looking at the log messages, this is not the same issue. I'd recommend reaching out to the mailing list or IRC to help get some advice on what might be wrong.

In the future, please avoid posting on closed issues, unless you are 100% it is the same issue and it is not fixed in master (or the latest rc). The information you've provided here can confuse others who come along trying to identify whether or not they are seeing this problem.

A simple google search for "Error running DeviceCreate (createSnapDevice) dm_task_run failed" yielded http://stackoverflow.com/questions/24709741/cant-run-docker-container-due-device-mapper-error, which looks like it may be related.

**Comment by zanella** [comment]
@stevvooe came from here: https://github.com/docker/docker/issues/9600

I don't participate in docker dev, I just downloaded the latest binary available at the docker site, so I really don't know if something which I don't actually know what is is already fixed on master :-)

But thanks for the "simple" search, even though it didn't fix it for me.

**Comment by stevvooe** [comment]
@zanella #9600 is not related either.

I am trying to give you advice so that you can actually find a resolution. If you don't know if it is fixed in master, then use one of the many other communication channels suggested to find out. Posting on closed tickets is both confusing and doesn't help to achieve your goal.

**Comment by stevvooe** [comment]
Re-opening since #13971 failed to address the root cause.

**Comment by packetcollision** [comment]
I have been able to reliably reproduce this issue by downloading multiple images with a common parent simultaneously via a bash loop in a local docker-machine VM:

This script reproduces it 100% of the time for me, using overlay or aufs (haven't tried devicemapper), with either vmware-fusion or virturalbox VMs.
```
#!/bin/bash
docker-machine create --driver vmwarefusion --vmwarefusion-memory-size 2048 --vmwarefusion-cpu-count 2 --engine-storage-driver overlay docker-9718
eval "$(docker-machine env docker-9718)"
for image in mesosphere/chronos:chronos-2.3.4-1.0.81.ubuntu1404-mesos-0.22.1-1.0.ubuntu1404 mesosphere/mesos-master:0.22.1-1.0.ubuntu1404 mesosphere/mesos-slave:0.22.1-1.0.ubuntu1404 mesosphere/marathon:v0.8.2-RC4 ; do
    docker pull $image &
done
wait
```

Docker info:
```
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Kernel Version: 4.0.9-boot2docker
Operating System: Boot2Docker 1.8.1 (TCL 6.3); master : 7f12e95 - Thu Aug 13 03:24:56 UTC 2015
CPUs: 2
Total Memory: 1.956 GiB
Name: docker-9718
ID: BQHH:QR3W:E2QZ:MXYK:F4AP:DIVW:SP3N:6TI4:4TNT:3V5P:EDCA:CCFC
Debug mode (server): true
Debug mode (client): false
Fds: 9
Goroutines: 16
EventsListeners: 0
Init Path: /usr/local/bin/docker
Docker Root Dir: /mnt/sda1/var/lib/docker
Registry: [https://index.docker.io/v1/]
Labels:
 provider=vmwarefusion
```

**Comment by LK4D4** [comment]
@packetcollision Just successfully run your script on master. Not all images were pulled from v2 though.

**Comment by packetcollision** [comment]
@LK4D4 Hmm, figures. Since it seems to be a timing thing, maybe download or computer speed makes a difference, although I was hoping it would be reliable since I was able to make it happen on two different macs at work and at home, as well as in a VirtualBox being run inside a VMware OS X install on OS X (I'm testing deploying a dev environment, in case you're wondering what on earth I'm doing nesting VMs). 

It's also worth noting that the error output from that script is pretty garbled by the progress updates from the other concurrent pulls. replacing the `docker pull $image &` line with `docker pull $image >/dev/null &` makes the errors obvious, but makes it harder to see what is going on when they are thrown.

**Comment by LK4D4** [comment]
@packetcollision haha, I don't know, I definitely made pull code little cleaner yesterday, not sure if it fixed some issues as side-effect.

**Comment by packetcollision** [comment]
@LK4D4 I could try testing on my machines with master. How do I go about getting master onto a VM provisioned with Docker Machine?

**Comment by pditommaso** [comment]
I'm experiencing a similar issue. See my comment here 
https://github.com/docker/docker/issues/15283#issuecomment-131112488

**Comment by mbentley** [comment]
@packetcollision - if you just need the binary, you can replace the existing one on your system from https://master.dockerproject.org

**Comment by packetcollision** [comment]
@LK4D4 Well I tried replacing the docker daemon in the VM with the one from the master builder and while I didn't get a hang, I *did* still get the same errors.

**Comment by spinolacastro** [comment]
I could reproduce the issue 100% in a fresh install running ext3 for /var/lib/docker and devicemapper.
You can have a look at stack trace here (http://www.fpaste.org/255689/14397699/)

[root@node1 ~]# docker info
Containers: 0
Images: 0
Storage Driver: devicemapper
 Pool Name: docker--vg-docker--pool
 Pool Blocksize: 524.3 kB
 Backing Filesystem: extfs
 Data file:
 Metadata file:
 Data Space Used: 19.4 MB
 Data Space Total: 25.73 GB
 Data Space Available: 25.71 GB
 Metadata Space Used: 53.25 kB
 Metadata Space Total: 46.14 MB
 Metadata Space Available: 46.08 MB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Library Version: 1.02.93-RHEL7 (2015-01-28)
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.10.0-123.8.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
CPUs: 4
Total Memory: 6.898 GiB
Name: node1.v3.ops.getupcloud.com
ID: XOCN:YWOM:YLPT:MO3G:ARX4:IPW2:Q5IZ:LUEE:3TAR:CHHS:CGGZ:3OVN

**Comment by gregarndt** [comment]
Just as a reference point, I am able to hit this using aufs and docker-1.9.0-dev.  I can capture any other information if it is helpful.

**Comment by stevvooe** [comment]
@gregarndt Thank you for the data point.

Is there any change you could try it out with the patch from #15728?

**Comment by gregarndt** [comment]
I certainly could give it a go.  Could you point me in the right direction with building a docker binary with this PR?

**Comment by euank** [comment]
@gregarndt 
You can check it out via running `git fetch origin pull/15728/head:pull_fix; git checkout pull_fix` in a clone of the docker repository and then build it with `make` per usual (assuming you don't need a dynamic binary for devicemapper or other reasons).

I'll throw in some info around the times I've seen this issue as well, though mine are mostly older and I don't have a solid repro either.
When I was running into it fairly frequently in 1.6.2, I got the [this stack trace](https://gist.github.com/euank/fb00bbfdd107196ca269).
I ran into something similar in 1.8.0-dev / commit c986f85 related to the error channel which was fixed by #15545. Stack trace for this can be found [here](https://gist.github.com/euank/d7803234ea47e65b195e).

The second stack trace definitely isn't relevant, and the first being 1.6.2 is probably not all that relevant, but it would be great to see others who run into this issue on recent dockers to run `sudo kill -SIGUSR1 $(pidof docker)` and post the stacktrace written to Docker's logfile.

**Comment by LK4D4** [comment]
@gregarndt We just merged, you can use master now.

**Comment by stevvooe** [comment]
@gregarndt You can get artifacts from https://master.dockerproject.org/.

Let us know if you need anything else!

**Comment by aaronlehmann** [comment]
I think there's a good chance this will be fixed by https://github.com/docker/docker/pull/15955

**Comment by stevvooe** [comment]
@aaronlehmann Was this not addressed by #15728?

**Comment by aaronlehmann** [comment]
It's possible that #15728 could solve these particular cases, but there are still outstanding issues that would manifest themselves as similar failures. #15955 attempts to solve these.

**Comment by skippy** [comment]
thanks folks.  fingers crossed on the patches.  This is hitting me pretty hard on 1.8.1 as I use the `alpine:3.2` as a common base image.  deleting that image and then pulling the stuck images works (as long as they are done serially).  Other info if needed:

```
$ docker info
Containers: 2
Images: 49
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Kernel Version: 4.0.5
Operating System: CoreOS 723.3.0
CPUs: 2
Total Memory: 3.863 GiB
Name: ip-10-0-11-133.us-west-2.compute.internal
ID: GQOC:A7Q3:2NEU:GFK6:QILO:VGQE:PSON:DFG6:DWHZ:3FMG:ZK2F:PFXQ
```

```
$ docker version
Client version: 1.6.2
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 7c8fca2-dirty
OS/Arch (client): linux/amd64
Server version: 1.6.2
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 7c8fca2-dirty
OS/Arch (server): linux/amd64
```

```
$ cat /etc/lsb-release
DISTRIB_ID=CoreOS
DISTRIB_RELEASE=723.3.0
DISTRIB_CODENAME="Red Dog"
DISTRIB_DESCRIPTION="CoreOS 723.3.0"
```

**Comment by dchen1107** [comment]
Our customers are still hitting the same issue with docker 1.8.2: https://github.com/kubernetes/kubernetes/issues/10623#issuecomment-140375510

**Comment by mboersma** [comment]
We are seeing the same issue consistently, and it's what has kept Deis on Docker 1.5.0, where apparently we will stay until this is fixed. (Nothing to do with devicemapper, BTW: this is overlayfs.)

Our product consists of several systemd units, each of which downloads its Docker image when started. Because we follow Docker best practices, the images share base layers in common. So multiple `docker pull` operations will immediately collide and corrupt the Docker graph.

What the user sees is a systemd unit that constantly restarts. This all worked just fine with Docker 1.5 / Registry v0.9.1. We can't upgrade to the new Go registry or newer versions of Docker because it will trigger this bug simply by deploying multiple app units with a `deis scale` operation.

**Comment by jmitchell** [comment]
Does this need to be assigned a new milestone? 1.7.1 is closed.

**Comment by LK4D4** [comment]
@jmitchell Yup, thank you. People reporting that it works in master, though.

**Comment by d3ming** [comment]
We are getting this issue fairly consistently with 1.8.2 whereas it used to work on an older version (I believe 1.6ish)

**Comment by pditommaso** [comment]
Same here. I confirm that `1.6.2` is the last working version.

**Comment by stdexcept** [comment]
This issue seriously affects us when new machines join our coreos cluster. This is definitely blocking us, as many others I can imagine.

**Comment by bitglue** [comment]
An expedient hack in many cases is to implement locking external to docker on any pulls or builds. `flock(1)` can accomplish this pretty easily, so wherever you have

    docker pull ...
    docker build ...

Then pick some file (doesn't matter what really, as long as it's the same everywhere) and change those commands to:

    flock /var/run/lock/docker docker pull ...
    flock /var/run/lock/docker docker build ...

**Comment by pditommaso** [comment]
@bitglue The Docker runtime should do that.

**Comment by bitglue** [comment]
@pditommaso you know what an expedient hack is, right?

**Comment by pditommaso** [comment]
@bitglue Um yes, but in my use case that hack is not an option.

**Comment by burke** [comment]
This was fixed by https://github.com/docker/docker/pull/15955 AFAICT.

**Comment by vbatts** [comment]
Burke, have checked this on master?

**Comment by burke** [comment]
Yep, I tried my repro from https://github.com/docker/docker/pull/16893 and it's fixed.

**Comment by vbatts** [comment]
Good to hear. Then docker-1.9 it is.

**Comment by icecrime** [comment]
Thanks @burke! :-)
